Can't exponentiate BigInts

In normal JavaScript, this works `BigInt(2) ** BigInt(2)`

But writing the above in TypeScript, compiles to `Math.pow(BigInt(2), BigInt(2))`, which doesn't work.

According to MDN:

> [BigInt] cannot be used with methods in the built-in Math object

---

TypeScript code

```ts
BigInt(2) ** BigInt(2)
```

**Expected behavior:**
Return 4.

**Actual behavior:**

```
TypeError: Cannot convert a BigInt value to a number
    at Math.pow (<anonymous>)
```

Hi @DanielRosenwasser ,

When we target es5 , we are getting "error TS2304: Cannot find name 'BigInt'.", if we change the error code to TS2737 which stands for "BigInt literals are not available when targeting lower than ESNext." it will fix the issue, let me know if we can change.

This looks like it was separately already solved. With this file:

```ts
3 ** 4n;
BigInt(2) ** BigInt(2);
```

Targeting `es5`, you get these errors:

```
index.ts:1:1 - error TS2365: Operator '**' cannot be applied to types '3' and '4n'.

1 3 ** 4n;
  ~~~~~~~

index.ts:1:6 - error TS2737: BigInt literals are not available when targeting lower than ESNext.

1 3 ** 4n;
       ~~

index.ts:2:1 - error TS2304: Cannot find name 'BigInt'.

2 BigInt(2) ** BigInt(2);
  ~~~~~~

index.ts:2:14 - error TS2304: Cannot find name 'BigInt'.

2 BigInt(2) ** BigInt(2);
               ~~~~~~
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -28440,6 +28440,12 @@ namespace ts {
                                 case SyntaxKind.GreaterThanGreaterThanGreaterThanToken:
                                 case SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken:
                                     reportOperatorError();
+                                    break;
+                                case SyntaxKind.AsteriskAsteriskToken:
+                                case SyntaxKind.AsteriskAsteriskEqualsToken:
+                                    if (languageVersion < ScriptTarget.ES2016) {
+                                        error(errorNode, Diagnostics.Exponentiation_cannot_be_performed_on_bigint_values_unless_the_target_option_is_set_to_es2016_or_later);
+                                    }
                             }
                             resultType = bigintType;
                         }

--- a/src/compiler/diagnosticMessages.json
+++ b/src/compiler/diagnosticMessages.json
@@ -2967,6 +2967,10 @@
         "category": "Error",
         "code": 2790
     },
+    "Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.": {
+        "category": "Error",
+        "code": 2791
+    },
 
     "Import declaration '{0}' is using private name '{1}'.": {
         "category": "Error",

--- a/tests/baselines/reference/bigIntWithTargetES2016.js
+++ b/tests/baselines/reference/bigIntWithTargetES2016.js
@@ -1,0 +1,11 @@
+//// [bigIntWithTargetES2016.ts]
+BigInt(1) ** BigInt(1); // should not error
+
+let num = BigInt(2);
+num **= BigInt(2); // should not error
+
+
+//// [bigIntWithTargetES2016.js]
+BigInt(1) ** BigInt(1); // should not error
+let num = BigInt(2);
+num **= BigInt(2); // should not error

--- a/tests/baselines/reference/bigIntWithTargetES2016.symbols
+++ b/tests/baselines/reference/bigIntWithTargetES2016.symbols
@@ -1,13 +1,13 @@
 === tests/cases/compiler/bigIntWithTargetES2016.ts ===
 BigInt(1) ** BigInt(1); // should not error
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
 
 let num = BigInt(2);
 >num : Symbol(num, Decl(bigIntWithTargetES2016.ts, 2, 3))
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
 
 num **= BigInt(2); // should not error
 >num : Symbol(num, Decl(bigIntWithTargetES2016.ts, 2, 3))
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
 

--- a/tests/baselines/reference/bigIntWithTargetES2016.symbols
+++ b/tests/baselines/reference/bigIntWithTargetES2016.symbols
@@ -1,0 +1,13 @@
+=== tests/cases/compiler/bigIntWithTargetES2016.ts ===
+BigInt(1) ** BigInt(1); // should not error
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+
+let num = BigInt(2);
+>num : Symbol(num, Decl(bigIntWithTargetES2016.ts, 2, 3))
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+
+num **= BigInt(2); // should not error
+>num : Symbol(num, Decl(bigIntWithTargetES2016.ts, 2, 3))
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+

--- a/tests/baselines/reference/bigIntWithTargetES2016.types
+++ b/tests/baselines/reference/bigIntWithTargetES2016.types
@@ -1,0 +1,23 @@
+=== tests/cases/compiler/bigIntWithTargetES2016.ts ===
+BigInt(1) ** BigInt(1); // should not error
+>BigInt(1) ** BigInt(1) : bigint
+>BigInt(1) : bigint
+>BigInt : BigIntConstructor
+>1 : 1
+>BigInt(1) : bigint
+>BigInt : BigIntConstructor
+>1 : 1
+
+let num = BigInt(2);
+>num : bigint
+>BigInt(2) : bigint
+>BigInt : BigIntConstructor
+>2 : 2
+
+num **= BigInt(2); // should not error
+>num **= BigInt(2) : bigint
+>num : bigint
+>BigInt(2) : bigint
+>BigInt : BigIntConstructor
+>2 : 2
+

--- a/tests/baselines/reference/bigIntWithTargetLessThanES2016.errors.txt
+++ b/tests/baselines/reference/bigIntWithTargetLessThanES2016.errors.txt
@@ -1,0 +1,14 @@
+tests/cases/compiler/bigIntWithTargetLessThanES2016.ts(1,1): error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+tests/cases/compiler/bigIntWithTargetLessThanES2016.ts(4,1): error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+
+
+==== tests/cases/compiler/bigIntWithTargetLessThanES2016.ts (2 errors) ====
+    BigInt(1) ** BigInt(1); // should error
+    ~~~~~~~~~~~~~~~~~~~~~~
+!!! error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+    
+    let foo = BigInt(2);
+    foo **= BigInt(2); // should error
+    ~~~~~~~~~~~~~~~~~
+!!! error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+    

--- a/tests/baselines/reference/bigIntWithTargetLessThanES2016.errors.txt
+++ b/tests/baselines/reference/bigIntWithTargetLessThanES2016.errors.txt
@@ -1,14 +1,14 @@
-tests/cases/compiler/bigIntWithTargetLessThanES2016.ts(1,1): error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
-tests/cases/compiler/bigIntWithTargetLessThanES2016.ts(4,1): error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+tests/cases/compiler/bigIntWithTargetLessThanES2016.ts(1,1): error TS2791: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+tests/cases/compiler/bigIntWithTargetLessThanES2016.ts(4,1): error TS2791: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
 
 
 ==== tests/cases/compiler/bigIntWithTargetLessThanES2016.ts (2 errors) ====
     BigInt(1) ** BigInt(1); // should error
     ~~~~~~~~~~~~~~~~~~~~~~
-!!! error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+!!! error TS2791: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
     
     let foo = BigInt(2);
     foo **= BigInt(2); // should error
     ~~~~~~~~~~~~~~~~~
-!!! error TS2782: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
+!!! error TS2791: Exponentiation cannot be performed on 'bigint' values unless the 'target' option is set to 'es2016' or later.
     

--- a/tests/baselines/reference/bigIntWithTargetLessThanES2016.js
+++ b/tests/baselines/reference/bigIntWithTargetLessThanES2016.js
@@ -1,0 +1,11 @@
+//// [bigIntWithTargetLessThanES2016.ts]
+BigInt(1) ** BigInt(1); // should error
+
+let foo = BigInt(2);
+foo **= BigInt(2); // should error
+
+
+//// [bigIntWithTargetLessThanES2016.js]
+Math.pow(BigInt(1), BigInt(1)); // should error
+let foo = BigInt(2);
+foo = Math.pow(foo, BigInt(2)); // should error

--- a/tests/baselines/reference/bigIntWithTargetLessThanES2016.symbols
+++ b/tests/baselines/reference/bigIntWithTargetLessThanES2016.symbols
@@ -1,13 +1,13 @@
 === tests/cases/compiler/bigIntWithTargetLessThanES2016.ts ===
 BigInt(1) ** BigInt(1); // should error
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
 
 let foo = BigInt(2);
 >foo : Symbol(foo, Decl(bigIntWithTargetLessThanES2016.ts, 2, 3))
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
 
 foo **= BigInt(2); // should error
 >foo : Symbol(foo, Decl(bigIntWithTargetLessThanES2016.ts, 2, 3))
->BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.es2020.bigint.d.ts, --, --), Decl(lib.es2020.bigint.d.ts, --, --))
 

--- a/tests/baselines/reference/bigIntWithTargetLessThanES2016.symbols
+++ b/tests/baselines/reference/bigIntWithTargetLessThanES2016.symbols
@@ -1,0 +1,13 @@
+=== tests/cases/compiler/bigIntWithTargetLessThanES2016.ts ===
+BigInt(1) ** BigInt(1); // should error
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+
+let foo = BigInt(2);
+>foo : Symbol(foo, Decl(bigIntWithTargetLessThanES2016.ts, 2, 3))
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+
+foo **= BigInt(2); // should error
+>foo : Symbol(foo, Decl(bigIntWithTargetLessThanES2016.ts, 2, 3))
+>BigInt : Symbol(BigInt, Decl(lib.esnext.bigint.d.ts, --, --), Decl(lib.esnext.bigint.d.ts, --, --))
+

--- a/tests/baselines/reference/bigIntWithTargetLessThanES2016.types
+++ b/tests/baselines/reference/bigIntWithTargetLessThanES2016.types
@@ -1,0 +1,23 @@
+=== tests/cases/compiler/bigIntWithTargetLessThanES2016.ts ===
+BigInt(1) ** BigInt(1); // should error
+>BigInt(1) ** BigInt(1) : bigint
+>BigInt(1) : bigint
+>BigInt : BigIntConstructor
+>1 : 1
+>BigInt(1) : bigint
+>BigInt : BigIntConstructor
+>1 : 1
+
+let foo = BigInt(2);
+>foo : bigint
+>BigInt(2) : bigint
+>BigInt : BigIntConstructor
+>2 : 2
+
+foo **= BigInt(2); // should error
+>foo **= BigInt(2) : bigint
+>foo : bigint
+>BigInt(2) : bigint
+>BigInt : BigIntConstructor
+>2 : 2
+

--- a/tests/cases/compiler/bigIntWithTargetES2016.ts
+++ b/tests/cases/compiler/bigIntWithTargetES2016.ts
@@ -1,0 +1,7 @@
+// @target: es2016
+// @lib: esnext
+
+BigInt(1) ** BigInt(1); // should not error
+
+let num = BigInt(2);
+num **= BigInt(2); // should not error

--- a/tests/cases/compiler/bigIntWithTargetLessThanES2016.ts
+++ b/tests/cases/compiler/bigIntWithTargetLessThanES2016.ts
@@ -1,0 +1,7 @@
+// @target: es2015
+// @lib: esnext
+
+BigInt(1) ** BigInt(1); // should error
+
+let foo = BigInt(2);
+foo **= BigInt(2); // should error

